x/provider/types/types.go

```diff
@@ -8,9 +8,29 @@ type Attributes []sdk.Attribute
 
 // Provider stores owner and host details
 type Provider struct {
-	Owner      sdk.AccAddress `json:"owner"`
-	HostURI    string         `json:"host-uri"`
-	Attributes Attributes     `json:"attributes"`
+	Owner         sdk.AccAddress `json:"owner"`
+	HostURI       string         `json:"host-uri"`
+	Attributes    Attributes     `json:"attributes"`
+	ReqAttributes Attributes     `json:"req-attributes" yaml:"req-attributes"`
+}
+
+// MatchReqAttributes method compares provided attributes with specific provider require attributes
+func (p Provider) MatchReqAttributes(attrs []sdk.Attribute) bool {
+loop:
+	for _, req := range p.ReqAttributes {
```

Comment:
Let's only do this in one place - move the current matching routine to a utility package if you have to.

There are assumptions baked into the existing algorithm and it's surprisingly easy to get things different.  `match(a,b)` isn't the same as `match(b,a)` for instance.